### PR TITLE
IMPRO1214 Bug fix to chunksize to improve speed of processing.

### PR DESCRIPTION
**Pull request by carolinejjones**
Description :  Bug fix to chunksize to improve speed of processing.

Testing:
 - [ ] Ran tests and they passed OK

### CODE REVIEW

**Comment by carolinejjones** [comment]
Tested failed improver-gradient 02-basic.bats

 ValueError: chunksize cannot exceed dimension size

This code is saving a cubelist rather than a single cube with different x-y dimensions.

File: lib/improver/utilities/save.py [hunk 1]
```diff
@@ -156,13 +156,14 @@ def save_netcdf(cubelist, filename):
...
-    if cube.ndim >= 2:
-        xy_chunksizes = [
-            min([128, cube.shape[-2]]),
-            min([128, cube.shape[-1]])
-        ]
-        chunksizes = tuple([1] * (cube.ndim - 2) + xy_chunksizes)
-    iris.fileformats.netcdf.save(cubelist, filename, local_keys=local_keys,
-                                 complevel=1, shuffle=True, zlib=True,
-                                 chunksizes=chunksizes)
+    if len(cubelist) > 1:
```
Comment (LaurenceBeard) [inline lib/improver/utilities/save.py]:
```suggestion
    if len(cubelist) > 2:
```

**Comment by LaurenceBeard** [comment]
An alternative suggestion, outside the loop:
```suggestion
        if len(set(cube.shape[-2:] for cube in cubelist)) == 1:
            chunksizes = tuple([1] * (cube.ndim - 2)) + cube.shape[-2:]
        else:
            chunksizes = None
            warn..
```
Which saves the need to check for ndims<2
But either solution I'm happy with, just the shuffle change requested.

Tested with spot data & gridded, seems to work either way.

**Comment by TomekTrzeciak** [comment]
@carolinejjones, @LaurenceBeard see also https://github.com/metoppv/improver/pull/876/commits/c191b8385130a1b0953a822c464cf5ffe5c4f220, which deals with the problem on the load, rather than save end (I've seen big speedup on stage files in a notebook, but haven't tried it HPC).

File: lib/improver/utilities/save.py [hunk 2]
```diff
@@ -137,15 +138,31 @@ def save_netcdf(cubelist, filename):
...
     """
     if isinstance(cubelist, iris.cube.Cube):
         cubelist = [cubelist]
 
+    chunksizes = None
+    xy_chunksizes = [None, None]
     for cube in cubelist:
         check_cube_not_float64(cube)
         _order_cell_methods(cube)
         _check_for_units(cube)
```
Comment (LaurenceBeard) [inline lib/improver/utilities/save.py]: Tomek wanted a comment to explain this, something like:
```suggestion
    # If all xy slices are the same shape, use this to determine the chunksize for the netCDF (eg. 1, 1, 970, 1042)
```

File: lib/improver/utilities/save.py [hunk 3]
```diff
@@ -137,14 +138,30 @@ def save_netcdf(cubelist, filename):
...
+    # If all xy slices are the same shape, use this to determine the chunksize for the netCDF (eg. 1, 1, 970, 1042)
```
Comment (LaurenceBeard) [inline lib/improver/utilities/save.py]:
```suggestion
    # If all xy slices are the same shape, use this to determine the
    # chunksize for the netCDF (eg. 1, 1, 970, 1042)
```

**Pull request merged**